js/viz.js

```diff
@@ -199,5 +204,56 @@ const viz = {
     const { width, height } = this.getDimensions('visualization');
     this.updateCanvas(width, height);
     this.draw(this.nodes, this.links);
+  },
+
+  addDrag() {
+    const drag = d3.drag();
+    drag.subject(() => this.dragSubject());
+    drag.on('end', () => this.dragEnded());
+
+    d3.select(this.canvas)
+      .call(drag);
+  },
+
+  dragSubject() {
+    const x = this.transform.invertX(d3.event.x);
+    const y = this.transform.invertY(d3.event.y);
+    const node = this.getNodeAtCoordinates(x, y);
```

Comment:
so glad that this piece of code which I wrote `getNodeAtCoordinates(x, y)` for tooltips is used here.